addons/account/models/account_move.py

```diff
@@ -1464,25 +1459,55 @@ def _track_subtype(self, init_values):
         # OVERRIDE to add custom subtype depending of the state.
         self.ensure_one()
 
-        if self.type not in ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt'):
+        if not self.is_invoice(include_receipts=True):
             return super(AccountMove, self)._track_subtype(init_values)
 
         if 'invoice_payment_state' in init_values and self.invoice_payment_state == 'paid':
             return self.env.ref('account.mt_invoice_paid')
-        elif 'state' in init_values and self.state == 'posted' and self.type in ('out_invoice', 'out_refund', 'out_receipt'):
+        elif 'state' in init_values and self.state == 'posted' and self.is_sale_document(include_receipts=True):
             return self.env.ref('account.mt_invoice_validated')
-        elif 'state' in init_values and self.state == 'draft' and self.type in ('out_invoice', 'out_refund', 'out_receipt'):
+        elif 'state' in init_values and self.state == 'draft' and self.is_sale_document(include_receipts=True):
             return self.env.ref('account.mt_invoice_created')
         return super(AccountMove, self)._track_subtype(init_values)
 
     # -------------------------------------------------------------------------
     # BUSINESS METHODS
     # -------------------------------------------------------------------------
 
-    @api.multi
-    def _is_invoice(self):
-        self.ensure_one()
-        return self.type in ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt')
+    @api.model
+    def _is_invoice(self, tested_type, include_receipts=False):
+        return tested_type in ['out_invoice', 'out_refund', 'in_refund', 'in_invoice'] + (include_receipts and ['out_receipt', 'in_receipt'] or [])
+
+    def is_invoice(self, include_receipts=False):
+        return self._is_invoice(self.type, include_receipts)
```

Comment:
why not:
```
@api.model
def get_invoice_types(self, include_receipts=False):
    return ['out_invoice', 'out_refund', 'in_refund', 'in_invoice'] + (include_receipts and ['out_receipt', 'in_receipt'] or [])

@api.multi
def is_invoice(self, include_receipts=False):
    return self.type in self.get_invoice_types(include_receipts=include_receipts)
```

that could be also be used as:

`move_type in self.env['account.move'].get_invoice_types()`